### DISCUSSION TITLE
AST to JSX conversion (?)

### DISCUSSION THREAD

**Discussion by rrjanbiah** [swc-project/swc#4241] [Q&A]
Context: I'm trying to create a React to SolidJS converter using swc-playground as a base https://github.com/rrjanbiah/react2solid/tree/main/converter With the base swc-playground, I'm able to convert the React code to AST. I want to do some manipulations in the AST and convert back to the original/JSX code.

Is AST to JSX/React code possible? If possible, please share relevant links.

If not possible, please suggest the starting point for the development.

TIA

**Comment by rrjanbiah** [comment]
_Replying my own question with what it is upto now_:

Babel has `generator` that is providing AST to JSX:

```js
import generate from "@babel/generator";
const { code } = generate(ast);
```

Though it looks promising, it doesn't seem to be working with SWC generated AST (Getting `unknown node of type "Module" with constructor "Object"` when tried to use)

**Comment by overlookmotel** [comment]
Babel's AST format and SWC's are different in quite a few ways. Also, some other tools use another format again called ESTree.

If you want to use `@babel/generator` for printing AST to code, you should do the whole thing in Babel - parse the input code in Babel so you get a Babel-format AST which then `@babel/generator` will understand.

On the other hand, if you want to use SWC, use SWC's `print()` or `printSync()` methods. They do exactly the same thing as `@babel/generator`, but with an SWC-format AST.

I think these methods are undocumented at present, but the code for them in SWC is here:

https://github.com/swc-project/swc/blob/9feabcd1452d7c05b89d8e6a3033fa8353c587e3/node-swc/src/index.ts#L92-L110

Hope that helps.

**Reply by rrjanbiah** [reply]
Looks like this is exactly I wanted. But, when I checked the swc web version, I'm getting `Cannot read properties of undefined (reading 'printSync')`. For now, I'm making a progress with the babel, but will definitely come back to swc little later... after getting a working version of my program.

**Comment by RiESAEX** [comment]
I think swc has codegen code for jsx..

**Reply by rrjanbiah** [reply]
`parseSync` is for JSX to AST. `transformSync` is for JSX to JavaScript. If you know any codegen for AST to JSX, please share.

**Comment by ecoinomist** [comment]
Any update on this? I'm also in need of an swc web version to transform AST to JSX output.

**Comment by AlecksJohannes** [comment]
Calling `swc.printSync` works for me. It prints out the code

**Reply by rrjanbiah** [reply]
Is that really working on swc web version?

**Comment by kwonoj** [comment]
Closing as adqueadtely answered, use parse / print pair as needed.
